fix: publish a release with draft relations

apfmelo opened the pull request:
### What does it do?
#### Fix at content-releases package
- refactor `getFormattedActions` to group actions by published or draft
- add `skipDraftFetching` at `getEntry`

### Why is it needed?

The publish release service wasn't publishing on the relation parent when the child was a draft (first time created)

### How to test it?

Using: https://github.com/strapi/strapi/tree/develop/examples/getstarted
Create a release with 2 actions: 
- Publishing a draft `kitchensink` that is associated with the draft `Tag`
- Publishing draft `Tag`

Publish the release
Expected behavior: the `kitchensink` published version must be at the `Tag` published version

### Related issue(s)/PR(s)

https://support.strapi.io/support/tickets/11803

## Updated

To handle related entries within the same release — especially those with one-way relationships — a tree structure is now used. Entries with no dependencies appear at the root level, while dependent entries are nested under their respective parent via a children list.

During publishing, the tree is traversed using a depth-first (DFS) strategy. Each entry is published asynchronously, ensuring that dependencies are respected and processed in the correct order.

```mermaid
graph TD
    subgraph "Root Level"
        8["ID: 8<br>contentType: api::tag.tag<br>documentId: ww104qpfnk1njkjbugxh2259<br>locale: en<br>type: publish<br>_depth: 0"]
        13["ID: 13<br>contentType: api::tag.tag<br>documentId: jwio3bitz25zbqz9iqik1rpo<br>locale: en<br>type: unpublish<br>_depth: 0"]
    end
    
    8 --> 40["ID: 40<br>contentType: api::kitchensink.kitchensink<br>documentId: fk3ppvivas06ovk2s3efkkzb<br>locale: en<br>type: publish<br>_depth: 1"]
    8 --> 9["ID: 9<br>contentType: api::tag.tag<br>documentId: pqtugnynmx5mvnhkgcaf4np8<br>locale: en<br>type: publish<br>_depth: 1"]
    
    40 --> 5["ID: 5<br>contentType: api::menu.menu<br>documentId: fzc3iofrkufxl75qcrjqkt8t<br>locale: null<br>type: publish<br>_depth: 2"]
    9 --> 1["ID: 1<br>contentType: api::homepage.homepage<br>documentId: eobknun5egzh3rmc7au0zerm<br>locale: en<br>type: publish<br>_depth: 2"]
    
    13 --> 15["ID: 15<br>contentType: api::tag.tag<br>documentId: tih3lm66pidqnssaqx5wf9kt<br>locale: en<br>type: unpublish<br>_depth: 1"]
    
    classDef publish fill:#d4f0c4,stroke:#333,stroke-width:1px,color:#000
    classDef unpublish fill:#ffcccc,stroke:#333,stroke-width:1px,color:#000
    
    class 8,40,9,5,1 publish
    class 13,15 unpublish
```

apfmelo committed:
fix: relation field search to filter the string using contains

apfmelo committed:
fix: publish release service

apfmelo committed:
fix: add console.log and fix the document id

apfmelo committed:
refactor: remove console.log and draft condition

apfmelo committed:
refactor: lint

davimdantas committed:
feat(content-releases): refactor getFormattedActions to group actions by published  or draft

davimdantas committed:
feat(content-releases): refactor getFormattedActions to group actions by published  or draft

davimdantas committed:
test(content-release release): mock strapi documents

davimdantas committed:
test(content-release): add strapi as argument to getEntryStatus

davimdantas committed:
test(content-release): validate that unpublished entries are been published before the entries with modifications

davimdantas committed:
refactor(content-release): reduce getEntry calls on getFormattedActions by creating param skipDraftFetching

apfmelo committed:
refactor: revert to strapi:develop

davimdantas committed:
Merge branch 'develop' into fix/publish-release

davimdantas committed:
Merge branch 'develop' into fix/publish-release

davimdantas committed:
Merge branch 'develop' into fix/publish-release

davimdantas committed:
Merge branch 'develop' into fix/publish-release

derrickmehaffy committed:
Merge branch 'develop' into fix/publish-release

davimdantas committed:
Merge branch 'develop' into fix/publish-release

derrickmehaffy commented:
New internal TID for this is 672

markkaylor commented:
Hi @apfmelo thanks for the PR! I'm just struggling to understand and reproduce the bug. Here's what I think the issue you are describing is:

1. Create a kitchensink document
2. Create a tag document
3. Attach tag to kitchensink to create a relation
4. Add both tag and kitchensink documents to a release with the action "publish"
5. Publish the release

Expected behavior: both documents should be published

I've done this and as far as I can tell it seems to be working. Would you be able to send a screen recording demonstrating the issue and how I can reproduce?

davimdantas commented:
https://github.com/user-attachments/assets/51bc4d42-297d-4448-84bd-0d07d0815358

DB to validate the error:
https://filebin.net/p1cf2cte7x37ofuc

davimdantas committed:
Merge branch 'develop' into fix/publish-release

davimdantas committed:
feat(releases): implement tree structure

To handle related entries within the same release — especially those with one-way relationships — a tree structure is now used. Entries with no dependencies appear at the root level, while dependent entries are nested under their respective parent via a children list.

During publishing, the tree is traversed using a depth-first (DFS) strategy. Each entry is published asynchronously, ensuring that dependencies are respected and processed in the correct order.

davimdantas committed:
test(releases): validate tree generation

davimdantas committed:
Merge branch 'develop' into fix/publish-release

davimdantas commented:
Hi @markkaylor , our previous solution was not complete. It did not handle one way relations.

We've updated this PR with a more thoughtful one

mathildeleg commented:
Hey there,
Thanks for the PR, I tested again based on your video, but on the latest version of Strapi, I do not encounter the issue. Since this PR is from last year, can you please confirm you still encounter the issue and maybe add more details on the reproduction steps to make sure I'm not missing anything?
Thanks so much 🙏🏻

davimdantas commented:
> Hey there, Thanks for the PR, I tested again based on your video, but on the latest version of Strapi, I do not encounter the issue. Since this PR is from last year, can you please confirm you still encounter the issue and maybe add more details on the reproduction steps to make sure I'm not missing anything? Thanks so much 🙏🏻

Sure, I will validate the same scenario in a new version

mathildeleg commented:
Hey @davimdantas have you been able to validate this scenario in a new version? 😊

mathildeleg commented:
Hey, closing this for now, feel free to re-open if you encounter the issue again and this PR fixes it :)
